### PR TITLE
Errcheck

### LINKED ISSUE #918: fix `errcheck` errors
Go through all the pieces in our codebase which don't properly check error values and make sure that they do. We'll probably have to exclude the autogenerated proto files.
Once we have it up to speed, we should probably add this to our tests or at least pre-commit hooks.

Here's the current list of complaints (excluding protos and tests):

```
22:05 $ make errcheck | grep -Ev '_test\.go:|\.pb\.go:'
errcheck -ignore='os:Close,net:Close,code.google.com/p/biogo.store/interval:.*,io:Write,bytes:Write.*'
./...
github.com/cockroachdb/cockroach/client/http_sender.go:162:23   defer resp.Body.Close()
github.com/cockroachdb/cockroach/client/txn.go:121:8    t.Run(Call{Args: etArgs, Reply: etReply})
github.com/cockroachdb/cockroach/gossip/gossip.go:516:24        util.RetryWithBackoff(retryOptions, func() (util.RetryStatus, error) {
github.com/cockroachdb/cockroach/gossip/infostore.go:89:15      is.visitInfos(func(g *group) error {
github.com/cockroachdb/cockroach/gossip/infostore.go:249:15     is.visitInfos(nil, func(i *info) error
{
github.com/cockroachdb/cockroach/gossip/infostore.go:351:18     delta.visitInfos(func(g *group) error {
github.com/cockroachdb/cockroach/gossip/infostore.go:355:20     is.registerGroup(gCopy)
github.com/cockroachdb/cockroach/gossip/infostore.go:385:15     is.visitInfos(func(g *group) error {
github.com/cockroachdb/cockroach/gossip/infostore.go:387:22     delta.registerGroup(gDelta)
github.com/cockroachdb/cockroach/gossip/infostore.go:391:17     delta.addInfo(i)
github.com/cockroachdb/cockroach/gossip/infostore.go:404:15     is.visitInfos(nil, func(i *info) error
{
github.com/cockroachdb/cockroach/gossip/infostore.go:420:15     is.visitInfos(nil, func(i *info) error
{
github.com/cockroachdb/cockroach/gossip/simulation/network.go:88:25     node.SetNodeDescriptor(&proto.NodeDescriptor{
github.com/cockroachdb/cockroach/gossip/simulation/network.go:99:16     node.AddInfo(gossip.KeyNodeCount, int64(nodeCount), time.Hour)
github.com/cockroachdb/cockroach/gossip/simulation/network.go:158:27    nodes[0].Gossip.AddInfo(gossip.KeySentinel, int64(cycle), time.Hour)
github.com/cockroachdb/cockroach/gossip/simulation/network.go:180:27    nodes[i].Gossip.AddInfo(nodes[i].Addr.String(), int64(cycle), time.Hour)
github.com/cockroachdb/cockroach/kv/db.go:122:20        defer r.Body.Close()
github.com/cockroachdb/cockroach/kv/db.go:148:9 w.Write(body)
github.com/cockroachdb/cockroach/kv/dist_sender.go:233:39       permMap.VisitPrefixesHierarchically(start, func(start, end proto.Key, config interface{}) (bool, error) {
github.com/cockroachdb/cockroach/kv/local_sender.go:148:19      store.ExecuteCmd(call.Args, call.Reply)
github.com/cockroachdb/cockroach/kv/rest.go:223:21      defer r.Body.Close()
github.com/cockroachdb/cockroach/kv/rest.go:253:20      defer r.Body.Close()
github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:324:23  tmpKV.RunTransaction(txnOpts, func(txn
*client.Txn) error {
github.com/cockroachdb/cockroach/main.go:44:23  flag.CommandLine.Set("alsologtostderr", "true")
github.com/cockroachdb/cockroach/multiraft/multiraft.go:650:23  s.multiNode.Campaign(context.Background(), groupID)
github.com/cockroachdb/cockroach/multiraft/transport.go:147:15  client.Close()
github.com/cockroachdb/cockroach/resource/embedded.go:24:10     gz.Close()
github.com/cockroachdb/cockroach/rpc/client.go:212:18   c.Client.Close()
github.com/cockroachdb/cockroach/rpc/server.go:130:15   go http.Serve(s.listener, s)
github.com/cockroachdb/cockroach/security/securitytest/embedded.go:26:10        gz.Close()
github.com/cockroachdb/cockroach/server/admin.go:175:20 defer r.Body.Close()
github.com/cockroachdb/cockroach/server/admin_client.go:39:23   defer resp.Body.Close()
github.com/cockroachdb/cockroach/server/cli/cli.go:62:10        w.Flush()
github.com/cockroachdb/cockroach/server/cli/start.go:231:17     server.SendQuit(Context)
github.com/cockroachdb/cockroach/server/node.go:411:23  n.lSender.VisitStores(func(s *storage.Store) error {
github.com/cockroachdb/cockroach/server/node.go:431:26  n.lSender.VisitStores(func(store *storage.Store) error {
github.com/cockroachdb/cockroach/server/server.go:121:21        s.kvDB.RegisterRPC(s.rpc)
github.com/cockroachdb/cockroach/server/server.go:242:22        w.WriteCloser.Close()
github.com/cockroachdb/cockroach/server/status.go:104:9 w.Write(b)
github.com/cockroachdb/cockroach/server/status.go:115:9 w.Write(b)
github.com/cockroachdb/cockroach/server/status.go:132:9 w.Write(b)
github.com/cockroachdb/cockroach/server/status.go:148:10        w.Write(buf[:length])
github.com/cockroachdb/cockroach/server/status.go:188:9 w.Write(b)
github.com/cockroachdb/cockroach/server/status.go:200:9 w.Write([]byte(`{"stores": []}`))
github.com/cockroachdb/cockroach/server/status.go:206:9 w.Write([]byte(`{"transactions": []}`))
github.com/cockroachdb/cockroach/server/testserver.go:166:29    ts.node.lSender.VisitStores(func(s *storage.Store) error {
github.com/cockroachdb/cockroach/storage/engine/engine.go:232:10        b.Clear(kv.Key)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:43:20   MVCCMergeRangeStat(engine, raftID, StatLiveBytes, ms.LiveBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:44:20   MVCCMergeRangeStat(engine, raftID, StatKeyBytes, ms.KeyBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:45:20   MVCCMergeRangeStat(engine, raftID, StatValBytes, ms.ValBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:46:20   MVCCMergeRangeStat(engine, raftID, StatIntentBytes, ms.IntentBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:47:20   MVCCMergeRangeStat(engine, raftID, StatLiveCount, ms.LiveCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:48:20   MVCCMergeRangeStat(engine, raftID, StatKeyCount, ms.KeyCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:49:20   MVCCMergeRangeStat(engine, raftID, StatValCount, ms.ValCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:50:20   MVCCMergeRangeStat(engine, raftID, StatIntentCount, ms.IntentCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:51:20   MVCCMergeRangeStat(engine, raftID, StatIntentAge, ms.IntentAge)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:52:20   MVCCMergeRangeStat(engine, raftID, StatGCBytesAge, ms.GCBytesAge)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:53:20   MVCCMergeRangeStat(engine, raftID, StatLastUpdateNanos, ms.LastUpdateNanos)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:58:18   MVCCSetRangeStat(engine, raftID, StatLiveBytes, ms.LiveBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:59:18   MVCCSetRangeStat(engine, raftID, StatKeyBytes, ms.KeyBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:60:18   MVCCSetRangeStat(engine, raftID, StatValBytes, ms.ValBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:61:18   MVCCSetRangeStat(engine, raftID, StatIntentBytes, ms.IntentBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:62:18   MVCCSetRangeStat(engine, raftID, StatLiveCount, ms.LiveCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:63:18   MVCCSetRangeStat(engine, raftID, StatKeyCount, ms.KeyCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:64:18   MVCCSetRangeStat(engine, raftID, StatValCount, ms.ValCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:65:18   MVCCSetRangeStat(engine, raftID, StatIntentCount, ms.IntentCount)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:66:18   MVCCSetRangeStat(engine, raftID, StatIntentAge, ms.IntentAge)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:67:18   MVCCSetRangeStat(engine, raftID, StatGCBytesAge, ms.GCBytesAge)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:68:18   MVCCSetRangeStat(engine, raftID, StatLastUpdateNanos, ms.LastUpdateNanos)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:726:17  engine.Clear(versionKey)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:868:14  engine.Merge(metaKey, data)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1085:16 engine.Clear(origKey)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1086:14 engine.Put(newKey, valBytes)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1104:14 engine.Clear(latestKey)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1116:15 engine.Clear(metaKey)
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1224:16 engine.Clear(iter.Key())
github.com/cockroachdb/cockroach/storage/engine/mvcc.go:1238:17 engine.Clear(iter.Key())
github.com/cockroachdb/cockroach/storage/range.go:698:19        go completionFunc()
github.com/cockroachdb/cockroach/storage/range_tree.go:161:10   tc.flush()
github.com/cockroachdb/cockroach/storage/split_queue.go:110:13  rng.AddCmd(&proto.AdminSplitRequest{
github.com/cockroachdb/cockroach/storage/store.go:458:19        s.multiraft.Start(s.stopper)
github.com/cockroachdb/cockroach/storage/store.go:618:22        s.ctx.Gossip.AddInfo(keyMaxCapacity, *storeDesc, ttlCapacityGossip)
github.com/cockroachdb/cockroach/util/cache.go:416:16   ic.tree.Insert(e, false)
github.com/cockroachdb/cockroach/util/cache.go:419:16   ic.tree.Delete(&entry{key: key}, false)
github.com/cockroachdb/cockroach/util/testing.go:78:14  os.RemoveAll(dir)
github.com/cockroachdb/cockroach/util/testing.go:84:15  os.RemoveAll(dir)
```

### PR DESCRIPTION
**Pull request by tamird**
Depends on #960.

Fixes #918.

### CODE REVIEW

File: client/kv_test.go [hunk 1]
```diff
@@ -135,8 +135,7 @@ func TestKVCommitReadOnlyTransaction(t *testing.T) {
 		calls = append(calls, call.Method())
 	}))
 	if err := client.RunTransaction(nil, func(txn *Txn) error {
-		txn.Run(Get(proto.Key("a")))
```
Comment (tschottdorf) [inline client/kv_test.go]: funny that we had code like that.

File: rpc/server.go [hunk 2]
```diff
@@ -127,7 +127,13 @@ func (s *Server) Listen() error {
 // listener.
 func (s *Server) Serve(handler http.Handler) {
 	s.handler = handler
-	go http.Serve(s.listener, s)
+	go func() {
+		if err := http.Serve(s.listener, s); err != nil {
+			// FIXME: uncommenting the line below break tests
```
Comment (tschottdorf) [inline rpc/server.go]: what goes wrong?

**Comment by tschottdorf** [comment]
Nice. LGTM :+1:

File: rpc/server.go (hunk 2 above)
Comment (tamird) [inline rpc/server.go]: I think certificate issues, can look again tomorrow

**Comment by tamird** [comment]
Well, there are some [interesting test failures](https://circle-artifacts.com/gh/cockroachdb/cockroach/2236/artifacts/0/tmp/circle-artifacts.q0qv9VQ/excerpt.txt) here that are non-obvious to me and I can't seem to reproduce them locally.

**Comment by tschottdorf** [comment]
the one in `TestRapid...` should be fixed if you just ignore the errors in any `CreateGroup` and `RemoveGroup` calls in this test (it's kosher).

**Comment by tschottdorf** [comment]
this one is likely a slow test machine:
`convergence_test.go:41: expected a fully-connected network within 8 cycles; took 11`

**Comment by tschottdorf** [comment]
and finally the infostore.. well, we'll need to look at those, but you made some changes in those locations, if it's not that the info is pretty verbose so I'm  sure we can nail it down.

**Comment by tamird** [comment]
@tschottdorf PTAL last commit

**Comment by tschottdorf** [comment]
LGTM

**Comment by tschottdorf** [comment]
though a `sync.WaitGroup` would've looked nicer.

**Comment by tamird** [comment]
I didn't know about `WaitGroup`, thanks! Fixed.

**Comment by tschottdorf** [comment]
other than that, LGTM :+1:

File: storage/range.go [hunk 3]
```diff
@@ -701,7 +701,10 @@ func (r *Range) addWriteCmd(args proto.Request, reply proto.Response, wait bool)
 	if wait {
 		return completionFunc()
 	}
-	go completionFunc()
+	go func() {
+		// error is lost forever :()
```
Comment (bdarnell) [inline storage/range.go]: It's not lost forever; it's logged at line 695. We should actually just move that block into this goroutine.

Comment (tamird) [inline storage/range.go]: good call, will do in the next PR

**Pull request merged**